src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ArrayConverter.cs

```diff
@@ -16,9 +17,9 @@ internal sealed class ArrayConverter<TCollection, TElement>
     {
         internal override bool CanHaveIdMetadata => false;
 
-        protected override void Add(TElement value, ref ReadStack state)
+        protected override void Add([AllowNull] TElement value, ref ReadStack state)
         {
-            ((List<TElement>)state.Current.ReturnValue!).Add(value);
+            ((List<TElement>)state.Current.ReturnValue!).Add(value!);
```

Comment:
I assume adding `!` to `value` is only necessary because the collection classes haven't been updated to add nullable annotationsy yet?

Reply:
They have actually: https://github.com/dotnet/runtime/issues/2339. Spoke with @buyaa-n offline and learned that the `T` for the collection determines if `null` can be passed, which is why `AllowNull` is not used.

If `T` is `string`, the add method does not expect to have `null` input, while if `T` is `string?`, `null` is okay. Callers can bang the compiler warning away and pass an input that could be `null` anyway, which we have to do since the input depends on what is in the JSON (which could be `null`).

Per related discussion in https://github.com/dotnet/runtime/pull/35826#discussion_r421109077, I'm reverting the change to the method signature here, so the bang doesn't happen here anymore, but in `IEnumerableDefautConverter` & `DictionaryDefaultConverter`, e.g. https://github.com/dotnet/runtime/blob/75845b5223c1a72ae6372638844eed4612990f40/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableDefaultConverter.cs#L70.

BTW, the reason for the nullability concern here is that the `Read` method of the generic converter has been annotated to possibly return `null`, so now we have to assert that passing the result of the read (which could be null) to the add method is okay.

Reply:
On a related note, there's an open issue for the serializer to prevent non-nullable reference types from being deserialized with null when "`#nullable`" is enabled in a project -https://github.com/dotnet/runtime/issues/1256 . Right now, we consider nullability metadata to be a compile-time/static analysis feature which doesn't affect the runtime behavior of the serializer. It seems that EF Core honors nullability annotations, so maybe we can consider this in the future.